cartographer/common/configuration_printer_main.cc

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2019 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include <vector>
+
+#include "absl/memory/memory.h"
+#include "absl/strings/str_split.h"
+#include "cartographer/common/configuration_file_resolver.h"
+#include "cartographer/common/lua_parameter_dictionary.h"
+#include "gflags/gflags.h"
+#include "glog/logging.h"
+
+DEFINE_string(configuration_directories, "",
+              "Comma separated list of directories in which configuration files"
+              " are searched, the last is always the Cartographer installation"
+              " to allow including files from there.");
+DEFINE_string(configuration_basename, "",
+              "Basename, i.e. not containing any directory prefix, of the "
+              "configuration file.");
+DEFINE_string(subdictionary, "",
+              "Only print a subdictionary referenced by its Lua ID, e.g.: "
+              "'--subdictionary trajectory_builder.trajectory_builder_3d'");
+
+namespace cartographer {
+namespace common {
+
+std::unique_ptr<LuaParameterDictionary> LoadLuaDictionary(
+    const std::vector<std::string>& configuration_directories,
+    const std::string& configuration_basename) {
+  auto file_resolver =
+      absl::make_unique<ConfigurationFileResolver>(configuration_directories);
+  const std::string code =
+      file_resolver->GetFileContentOrDie(configuration_basename);
+  // We use no reference count because we just want to print the configuration.
+  return LuaParameterDictionary::NonReferenceCounted(code,
+                                                     std::move(file_resolver));
+}
+
+void PrintSubdictionaryById(LuaParameterDictionary* lua_parameter_dictionary,
```

Comment:
Is this really a feature people will use?
If unsure, we usually err on the side of not adding a new feature, so I vote for dropping this function.

Reply:
I added it for convenience to restrict output for readability (currently 230 lines total, duplicated parameter names across subdicts) if one is only interested only in a part. For example, for comparing some `*_2d` part when I don't care about the corresponding `*_3d` counterpart, which would also pop up in `grep` if the basenames are the same.

Probably not the killer feature everyone waited for, but so is probably this whole executable. But also doesn't hurt to have one more function IMO.